File: python/src/llikert/checkpoint.py
"""Local checkpoint directories (checkpoint schema v1, docs/checkpoint-format.md).

    run-dir/
      manifest.json          run identity: prepared task, engine, ordered ids, text hashes
      .lock/owner.json       present while a writer is active (created atomically with mkdir)
      chunks/chunk-000001.json
"""

from __future__ import annotations

import json
import os
import pathlib
import shutil
import socket
from typing import Any

from llikert._errors import CheckpointError
from llikert.result import RETRYABLE_STATUSES, ScoreResult, client_block, utc_now, write_json_atomic

CHECKPOINT_SCHEMA_VERSION = 1
PENDING_STATUS = "pending"


def _read_json(path: pathlib.Path) -> Any:
    try:
        with path.open(encoding="utf-8") as f:
            return json.load(f)
    except (OSError, ValueError) as exc:
        raise CheckpointError(f"cannot read {path.name}: {type(exc).__name__}") from None


def merge_chunks(chunk_dir: pathlib.Path) -> dict[str, dict[str, Any]]:
    """Merge committed chunks in order. A later record replaces an earlier one only when the
    earlier status is retryable; successes and deterministic input errors are final."""
    merged: dict[str, dict[str, Any]] = {}
    if not chunk_dir.is_dir():
        return merged
    for path in sorted(chunk_dir.glob("chunk-*.json")):
        chunk = _read_json(path)
        if chunk.get("checkpoint_schema_version") != CHECKPOINT_SCHEMA_VERSION:
            raise CheckpointError(f"{path.name} has an unsupported schema version")
        for record in chunk["results"]:
            existing = merged.get(record["id"])
            if existing is None or existing["status"] in RETRYABLE_STATUSES:
                merged[record["id"]] = record
    return merged


class Checkpoint:
    def __init__(self, path: pathlib.Path, manifest: dict[str, Any]):
        self.path = path
        self.manifest = manifest
        self.completed = merge_chunks(path / "chunks")
        self._next = 1 + max((int(p.stem.split("-")[1]) for p in (path / "chunks").glob("chunk-*.json")), default=0)
        self._locked = False

    # -- lifecycle ------------------------------------------------------------------------
    @classmethod
    def open(
        cls,
        path: str | os.PathLike,
        *,
        prepared: dict[str, Any],
        engine_info: dict[str, Any],
        ids: list[str],
        text_hashes: list[str | None],
        resume: bool,
        force_unlock: bool = False,
    ) -> Checkpoint:
        path = pathlib.Path(path)
        manifest_path = path / "manifest.json"
        expected = {
            "prepared_hash": prepared["prepared_hash"],
            "engine_fingerprint": prepared["engine_fingerprint"],
            "ids": ids,
            "text_sha256": text_hashes,
        }
        if manifest_path.exists():
            if not resume:
                raise CheckpointError(f"checkpoint {path} already exists; pass resume=True to continue it or choose a new path")
            manifest = _read_json(manifest_path)
            if manifest.get("checkpoint_schema_version") != CHECKPOINT_SCHEMA_VERSION:
                raise CheckpointError("the checkpoint has an unsupported schema version")
            found = {
                "prepared_hash": manifest["prepared"]["prepared_hash"],
                "engine_fingerprint": manifest["engine_fingerprint"],
                "ids": manifest["ids"],
                "text_sha256": manifest["text_sha256"],
            }
            changed = [k for k in expected if expected[k] != found[k]]
            if changed:
                what = {"prepared_hash": "task or preparation", "engine_fingerprint": "model or execution configuration",
                        "ids": "item ids or their order", "text_sha256": "texts"}
                raise CheckpointError(
                    "the checkpoint belongs to a different run (changed: " + ", ".join(what[k] for k in changed) + "); start a new checkpoint"
                )
            checkpoint = cls(path, manifest)
        else:
            if path.exists() and any(path.iterdir()):
                raise CheckpointError(f"{path} exists and is not a llikert checkpoint")
            (path / "chunks").mkdir(parents=True, exist_ok=True)
            manifest = {
                "checkpoint_schema_version": CHECKPOINT_SCHEMA_VERSION,
                "created_at": utc_now(),
                "client": client_block(),
                "protocol_version": 1,
                "engine_fingerprint": prepared["engine_fingerprint"],
                "engine": engine_info["engine"],
                "execution": engine_info.get("execution", {}),
                "prepared": prepared,
                "ids": ids,
                "text_sha256": text_hashes,
            }
            write_json_atomic(manifest_path, manifest)
            checkpoint = cls(path, manifest)
        checkpoint._lock(force_unlock)
        return checkpoint

    def _lock(self, force: bool) -> None:
        lock = self.path / ".lock"
        if force and lock.exists():
            shutil.rmtree(lock)
        try:
            lock.mkdir()
        except FileExistsError:
            owner = {}
            try:
                owner = json.loads((lock / "owner.json").read_text())
            except (OSError, ValueError):
                pass
            raise CheckpointError(
                f"checkpoint is locked by another writer (host {owner.get('host')}, pid {owner.get('pid')}, since {owner.get('created_at')}); "
                "if that run is no longer active, pass force_unlock=True"
            ) from None
        (lock / "owner.json").write_text(json.dumps({"host": socket.gethostname(), "pid": os.getpid(), "created_at": utc_now(), "client": client_block()}))
        self._locked = True

    def release(self) -> None:
        if self._locked:
            shutil.rmtree(self.path / ".lock", ignore_errors=True)
            self._locked = False

    # -- writing --------------------------------------------------------------------------
    def commit(self, records: list[dict[str, Any]]) -> None:
        if not records:
            return
        chunk = {
            "checkpoint_schema_version": CHECKPOINT_SCHEMA_VERSION,
            "sequence": self._next,
            "engine_fingerprint": self.manifest["engine_fingerprint"],
            "prepared_hash": self.manifest["prepared"]["prepared_hash"],
            "results": records,
        }
        write_json_atomic(self.path / "chunks" / f"chunk-{self._next:06d}.json", chunk)
        self._next += 1
        for record in records:
            existing = self.completed.get(record["id"])
            if existing is None or existing["status"] in RETRYABLE_STATUSES:
                self.completed[record["id"]] = record

    # -- reading ----------------------------------------------------------------------------
    def to_result(self, allow_incomplete: bool = False) -> ScoreResult:
        return result_from_manifest(self.manifest, self.completed, allow_incomplete)


def result_from_manifest(manifest: dict[str, Any], completed: dict[str, dict[str, Any]], allow_incomplete: bool) -> ScoreResult:
    prepared = manifest["prepared"]
    k = len(prepared["categories"])
    numeric = prepared["categories"][0]["value"] is not None
    items = []
    for item_id, sha in zip(manifest["ids"], manifest["text_sha256"]):
        record = completed.get(item_id)
        if record is None:
            if not allow_incomplete:
                raise CheckpointError("the checkpoint is incomplete; resume it, or pass allow_incomplete=True")
            record = {
                "id": item_id, "status": PENDING_STATUS, "error": None,
                "candidate_log_probs": [None] * k, "candidate_probs": [None] * k, "probabilities": [None] * k,
                "log_coverage": None, "coverage": None, "n_prompt_tokens": None, "prompt_token_sha256": None,
                "warnings": [], **({"expected_value": None} if numeric else {}), "text_sha256": sha,
            }
        items.append(record)
    return ScoreResult(
        prepared=prepared,
        engine_fingerprint=manifest["engine_fingerprint"],
        engine=manifest["engine"],
        execution=manifest.get("execution", {}),
        items=items,
        created_at=utc_now(),
        client=client_block(),
    )


def read_checkpoint(path: str | os.PathLike, allow_incomplete: bool = False) -> ScoreResult:
    """Read a checkpoint directory offline, without contacting a service."""
    path = pathlib.Path(path)
    manifest = _read_json(path / "manifest.json")
    if manifest.get("checkpoint_schema_version") != CHECKPOINT_SCHEMA_VERSION:
        raise CheckpointError("the checkpoint has an unsupported schema version")
    return result_from_manifest(manifest, merge_chunks(path / "chunks"), allow_incomplete)

File: r/llikert/R/checkpoint.R
# Local checkpoint directories (checkpoint schema v1, docs/checkpoint-format.md).
#
#   run-dir/manifest.json          run identity: prepared task, engine, ordered ids, text hashes
#   run-dir/.lock/owner.json       present while a writer is active (created atomically by dir.create)
#   run-dir/chunks/chunk-000001.json

merge_chunks <- function(chunk_dir) {
  merged <- list()
  files <- sort(list.files(chunk_dir, pattern = "^chunk-[0-9]+\\.json$", full.names = TRUE))
  for (path in files) {
    chunk <- tryCatch(read_json_file(path), error = function(e) llikert_abort(sprintf("Cannot read %s.", basename(path)), "checkpoint"))
    if (!identical(as.integer(chunk$checkpoint_schema_version), checkpoint_schema_version)) {
      llikert_abort(sprintf("%s has an unsupported schema version.", basename(path)), "checkpoint")
    }
    for (record in chunk$results) {
      existing <- merged[[record$id]]
      if (is.null(existing) || existing$status %in% retryable_statuses) merged[[record$id]] <- record
    }
  }
  merged
}

checkpoint_open <- function(path, prepared, engine, ids, hashes, resume, force_unlock, call = rlang::caller_env()) {
  manifest_path <- file.path(path, "manifest.json")
  if (file.exists(manifest_path)) {
    if (!resume) {
      llikert_abort(sprintf("Checkpoint %s already exists; use `resume = TRUE` to continue it or choose a new path.", path), "checkpoint", call = call)
    }
    manifest <- read_json_file(manifest_path)
    if (!identical(as.integer(manifest$checkpoint_schema_version), checkpoint_schema_version)) {
      llikert_abort("The checkpoint has an unsupported schema version.", "checkpoint", call = call)
    }
    changed <- c(
      "task or preparation" = !identical(manifest$prepared$prepared_hash, prepared$prepared_hash),
      "model or execution configuration" = !identical(manifest$engine_fingerprint, prepared$engine_fingerprint),
      "item ids or their order" = !identical(unlist(manifest$ids), ids),
      "texts" = !identical(manifest$text_sha256, hashes)
    )
    if (any(changed)) {
      llikert_abort(
        sprintf("The checkpoint belongs to a different run (changed: %s); start a new checkpoint.", paste(names(changed)[changed], collapse = ", ")),
        "checkpoint", call = call
      )
    }
  } else {
    if (dir.exists(path) && length(list.files(path, all.files = TRUE, no.. = TRUE))) {
      llikert_abort(sprintf("%s exists and is not a llikert checkpoint.", path), "checkpoint", call = call)
    }
    dir.create(file.path(path, "chunks"), recursive = TRUE, showWarnings = FALSE)
    manifest <- list(
      checkpoint_schema_version = checkpoint_schema_version,
      created_at = utc_now(),
      client = client_block(),
      protocol_version = protocol_version,
      engine_fingerprint = prepared$engine_fingerprint,
      engine = engine$info$engine,
      execution = engine$info$execution %||% list(),
      prepared = prepared,
      ids = as.list(ids),
      text_sha256 = hashes
    )
    write_json_atomic(manifest, manifest_path)
  }
  lock_checkpoint(path, force_unlock, call)
  chunk_files <- list.files(file.path(path, "chunks"), pattern = "^chunk-[0-9]+\\.json$")
  sequence <- if (length(chunk_files)) max(as.integer(sub("^chunk-([0-9]+)\\.json$", "\\1", chunk_files))) else 0L
  store <- new.env(parent = emptyenv())
  store$path <- path
  store$manifest <- manifest
  store$completed <- merge_chunks(file.path(path, "chunks"))
  store$sequence <- sequence
  store
}

lock_checkpoint <- function(path, force, call) {
  lock <- file.path(path, ".lock")
  if (force && dir.exists(lock)) unlink(lock, recursive = TRUE)
  if (!suppressWarnings(dir.create(lock))) {
    owner <- tryCatch(read_json_file(file.path(lock, "owner.json")), error = function(e) list())
    llikert_abort(
      sprintf("The checkpoint is locked by another writer (host %s, pid %s, since %s); if that run is no longer active, use `force_unlock = TRUE`.",
              owner$host %||% "?", owner$pid %||% "?", owner$created_at %||% "?"),
      "checkpoint", call = call
    )
  }
  writeLines(as.character(to_json(list(host = Sys.info()[["nodename"]], pid = Sys.getpid(), created_at = utc_now(), client = client_block()))),
             file.path(lock, "owner.json"))
}

checkpoint_release <- function(store) {
  unlink(file.path(store$path, ".lock"), recursive = TRUE)
}

checkpoint_commit <- function(store, records) {
  if (!length(records)) return(invisible())
  store$sequence <- store$sequence + 1L
  chunk <- list(
    checkpoint_schema_version = checkpoint_schema_version,
    sequence = store$sequence,
    engine_fingerprint = store$manifest$engine_fingerprint,
    prepared_hash = store$manifest$prepared$prepared_hash,
    results = unname(records)
  )
  write_json_atomic(chunk, file.path(store$path, "chunks", sprintf("chunk-%06d.json", store$sequence)), pretty = FALSE)
  for (record in records) {
    existing <- store$completed[[record$id]]
    if (is.null(existing) || existing$status %in% retryable_statuses) store$completed[[record$id]] <- record
  }
  invisible()
}

#' Read a checkpoint directory offline
#'
#' Builds a result from a checkpoint without contacting a service.
#'
#' @param path Checkpoint directory given to [score_items()].
#' @param allow_incomplete If `TRUE`, items not yet scored appear with status
#'   `"pending"`; otherwise an incomplete checkpoint is an error.
#' @return An `llikert_result`.
#' @export
read_checkpoint <- function(path, allow_incomplete = FALSE) {
  manifest_path <- file.path(path, "manifest.json")
  if (!file.exists(manifest_path)) llikert_abort(sprintf("%s is not a llikert checkpoint.", path), "checkpoint")
  manifest <- read_json_file(manifest_path)
  if (!identical(as.integer(manifest$checkpoint_schema_version), checkpoint_schema_version)) {
    llikert_abort("The checkpoint has an unsupported schema version.", "checkpoint")
  }
  completed <- merge_chunks(file.path(path, "chunks"))
  result_from_manifest(manifest, completed, allow_incomplete)
}

result_from_manifest <- function(manifest, completed, allow_incomplete, call = rlang::caller_env()) {
  prepared <- manifest$prepared
  k <- length(prepared$categories)
  numeric <- !is.null(prepared$categories[[1]]$value)
  ids <- unlist(manifest$ids)
  records <- lapply(seq_along(ids), function(i) {
    record <- completed[[ids[[i]]]]
    if (is.null(record)) {
      if (!allow_incomplete) llikert_abort("The checkpoint is incomplete; resume it, or use `allow_incomplete = TRUE`.", "checkpoint", call = call)
      record <- pending_record(ids[[i]], k, numeric, manifest$text_sha256[[i]])
    }
    record
  })
  new_result(prepared, manifest$engine_fingerprint, manifest$engine, manifest$execution %||% list(), records)
}
